### ISSUE TITLE
Background process keeps killing itself on Manjaro

### ISSUE THREAD

**Issue by aorozsa** [wwmm/pulseeffects#717]
As the title says, the background process created by this command: `pulseeffects --gapplication-service` (found in the automatically generated start on login config) keeps ending at random times on my Manjaro system. I know the process is dead because, for starters, my audio stops working, and running `killall pulseeffects` returns this: `pulseeffects: no process found`. And after restarting it, sound works again.

Precisely when it decides to keel over is random. I've witnessed it happen after waking from sleep, and sometimes I don't. I've witnessed it happen when I open Firefox for example, and sometimes I don't. And I've witnessed it happen after I was away from my computer for some time. The number of times it happens is less than those it doesn't, but it's still really annoying to have to keep on restarting it. And one more thing: even after restarting it with these commands: 
`pulseeffects --gapplication-service &`
`disown`
, it still likes to kill itself.

I'm using KDE and its version is 5.18.5, and the version of pulseeffects is 4.7.2. I would submit logs, but I have no idea what exactly to submit. Searched my system but couldn't find anything.

**Comment by wwmm** [comment]
First let's try to see some logs. Kill the current instance `pulseeffects -q` and restart it in debug mode `G_MESSAGES_DEBUG=pulseeffects pulseeffects`. This should print lots of things in the terminal.

It is clear that a segmentation fault is happening. The question is why. Maybe one of PulseEffects dependencies needs to be updated. In the next time you see this crash happening try to run the command `sudo coredumpctl info` as soon as possible. Maybe it shows something interesting. But as Arch Linux(and possible Manjaro) packages are not built with debug symbols it is possible nothing useful is shown.

**Comment by aorozsa** [comment]
@wwmm I'd like to clarify something. So, running `G_MESSAGES_DEBUG=pulseeffects pulseeffects` opens the pulseeffects window instead of running the service in the background like the command in my last post does. Do I keep this window open until **it** crashes and then run `sudo coredumpctl info`, or do I run the former command, copy-paste what comes up in the terminal, then close it, run the background service command, then run `sudo coredumpctl info` when **that** crashes?

**Comment by wwmm** [comment]
The second option. First let's see if there is anything strange in the logs. Then run it as usual and wait for a crash.

**Comment by aorozsa** [comment]
@wwmm Here's the output of [G_MESSAGES_DEBUG=pulseeffects pulseeffects](https://paste.ubuntu.com/p/kW5JhK472n/)

I'm now running the background service and waiting for that to crash.

**Comment by Digitalone1** [comment]
Presets manager is destroyed with no reason. Maybe a corrupted preset is causing the issue and that's possible since it's a know bug of 4.7.2 version. Why Manjaro hasn't updated to 4.7.3 yet?

Try to delete all presets or build pulseeffects-git from AUR.

**Comment by aorozsa** [comment]
So after deleting my profile (which I've been using for nearly two years) deleting `~/.config/PulseEffects` and resetting the settings, I still got `presets_manager: destroyed` upon starting and stopping pulseeffects, as shown [here](https://paste.ubuntu.com/p/sZbn5NMqYh/).

I uninstalled pulseeffects and built pulseeffects-git from the AUR, and running the command in the above pastebin yielded the same result, albeit with different plugins missing.

Manjaro's stable branch takes time to update because a bunch of testing is done on the software before it's released to the stable branch.

**Comment by Digitalone1** [comment]
A previous user resolved a weird issue resetting PE config and Pulseaudio local config (deleting `~/.config/pulse`).

You can try that too.

**Comment by aorozsa** [comment]
Still getting `presets_manager: destroyed`. I'll see if the background service crashes tomorrow.

**Comment by Digitalone1** [comment]
It's not a presets manager issue. It has been destroyed even on my system upon the start, I just check it out, my mistake. Anyway your logs does not show anything relevant and resetting PE and PA config haven't helped.

Let's wait @wwmm, maybe a look at pulseaudio log can give some hints.

**Comment by wwmm** [comment]
This log is fine. But unless I missed something you did not try to play something. Maybe the crash is happening only when PulseEffets pipeline alternates between the playing and paused state. IF that is the case maybe something interesting will be shown at this moment.

Usually we do not have segmentation faults when Pulseaudio does something odd. Something else is going on.

**Comment by aorozsa** [comment]
Wait a minute... I think I know what might be going on. I'm running Folding@Home. I'll finish up the current WUs I have, stop FAH for some time and observe the behaviour of pulseeffects.

**Comment by wwmm** [comment]
What is `Folding@Home`?

**Comment by aorozsa** [comment]
@wwmm It's a distributed computing project that gets everyday people to use their computers to simulate protein folding. You set up the client (in the case of Arch distros, the AUR package and its dependencies are [here](https://aur.archlinux.org/packages/foldingathome/)), then you're sent Work Units (WUs), your computer computes them, and then they're sent back when they're done. The reason I suggested it might be the problem is because my CPU is constantly at maximum usage (although I can do light things such as browse the internet and play light games just fine with FAH running) and that might've caused pulseeffects to stop (?). Oh, and my CPU and motherboard are old. 2600k & GA-Z68X-UD3H-B3.

**Comment by Digitalone1** [comment]
It's possible. I have issues on an old laptop where sometimes PulseEffects does not output audio.

**Comment by wwmm** [comment]
At first sight I can see high cpu usage causing stuttering. But not segmentation fault. Unless the system is deciding to kill our process when under pressure. But in this case why only us? 

Is the option `Priority Type` in our settings menu configured to `Real Time`?

**Comment by aorozsa** [comment]
I have it currently set to that. Today I'll test to see if having it set to Real Time and running FAH causes it to crash, since I still have active WUs. If I get a result early, I'll then proceed to see what happens when FAH isn't running (though that'll probably take a week to get the full picture).

**Comment by aorozsa** [comment]
Update: so far I've experienced two crashes during normal usage while FAH was running on the same day, one of them being right after The Elder Scrolls IV: Oblivion crashed and the other one being after I safely closed Spotify (adblock) (from AUR). Last night I left my computer on with FAH running (kind of regret it; got barely any sleep), and the pulseeffects service is still running and working properly. My guess to what exactly is happening is that pulseeffects will only **sometimes** crash while CPU usage is high, sound is being actively outputted and the application using outputting said sound suddenly ends, either by crash or closure. Today I'll begin testing to see if high CPU usage is the problem. In FAH, I'll pause CPU folding but allow GPU folding to continue. If I experience a pulseeffects crash any time over the next week, I'll also disable GPU folding. If that test also prematurely ends, then my guess is that my ageing motherboard's built-in audio is at the end of the road.

**Comment by aorozsa** [comment]
Nope! I encountered one even while I wasn't doing any folding at all.

**Comment by wwmm** [comment]
I wonder if the system logs have something that could help. Try run the command `sudo journalctl -b`. It will show the log of the current boot. In case you have restarted you can run `sudo journalctl -b -1` to get the log of the previous boot. Replace `1` by higher numbers in case it is necessary to go further in the past.

**Comment by aorozsa** [comment]
@wwmm Are there any specific parts of it you need that I can filter out with grep? It's an incredibly long log. Not even Ubuntu Pastebin can handle it.

**Comment by wwmm** [comment]
First I would search for lines with the word `PulseEffects` on them. Then I would copy them and maybe the 5 or 10 lines that come before and after.

**Comment by aorozsa** [comment]
This is the output of [sudo journalctl -b -1 | grep "pulseeffects" -A 10 -B 10](https://paste.ubuntu.com/p/Ck3dV72GS5/). Searching for "PulseEffects" didn't give anything, and the only two occurrences of "pulseeffects" are the lines that include "pulseeffects-service.desktop". The rest of the stuff doesn't seem like it's useful (at least to me).

Now that I think about it I never had this problem on Pop!_OS (which is an Ubuntu-based distro which uses Gnome & GTK). From what I've seen, KDE is significantly different from the stuff that distros utilising Gnome use. Do you think it'd be worth to back up my system and then install Manjaro Gnome to test if the DE is the culprit? I'm not comfortable with installing another DE on top of my current installation because, I've done it before on Pop!_OS, and stuff breaks.

**Comment by Digitalone1** [comment]
Give it a try, but I don't think that would be of any help. I used KDE, then switched to Gnome, but on both I'm getting no audio from PulseEffects  when the system is under heavy load. Could be an issue not specific to PulseEffects, but Pulseaudio or Gstreamer.

**Comment by aorozsa** [comment]
I encountered a crash in Manjaro Gnome. KDE is not the culprit.
I started watching a youtube video for a minute paused it, left for a couple of minutes and came back to find pulseeffects dead. @wwmm Take a look at the below. Are the last two lines of any use? They didn't occur at the time I was doing the above.

[the below](https://paste.ubuntu.com/p/pFftpNQ4yd/)

**Comment by Digitalone1** [comment]
They mean PulseEffects can't process those two apps. Something is preventing them to do so or they specifically won't allow other process to redirect their streams.

**Comment by wwmm** [comment]
Pulseaudio allows app developers to forbid custom routing. That is why these errors are happening. We are forbid to move them to another device. But this should not case a crash. I use Steam frequently with no issues.

**Comment by wwmm** [comment]
What is odd is that your Manjaro installation should not be much different from the one I have on Arch Linux. I am without ideas about why this random crash happens. Let's try another approach. Run PulseEffects always in debug mode for a while. It will help to see what is the last line printed when it crashes.

**Comment by aorozsa** [comment]
@wwmm I ran pulseeffects in debug mode for the whole day until now. I had a read through the log, and it seems pretty **normal. Except the last line...**

[the log](https://paste.ubuntu.com/p/G4ng9gFbpc/)

**Comment by wwmm** [comment]
The kernel is killing us. Probably its "out of memory killer". I think I might know what is going on. Do you open and close PulseEffects window many times along the day? When I as preparing PE `4.7.3` I noticed a very bad memory leak that happened whenever PE was running as a service and its window was closed and opened again. This was happening because I used a shared smart pointer in a place that made GTK think that it should not free the window memory because there was still a reference somewhere. As you are still on `4.7.2` you have this leak.

This would explain why this is so random but PE is always the one killed. The kernel "out of memory killer" kills the process using more memory. And because of the leak this is probably PE.

**Comment by wwmm** [comment]
If that is the cause of the crash updating to `4.7.3` or using the aur pkgbuild  `pulseeffects-git` should fix it.

**Comment by aorozsa** [comment]
@wwmm No. Under normal usage (which is starting it on login), it starts in the background and then I go about my day. I did the same thing yesterday when gathering the log: ran the command once, and didn't touch it.

I'll try testing pulseeffects-git. Just to make sure, what leftover files/folders should I delete (i.e ~/.config/PulseEffects/) before upgrading?

**Comment by aorozsa** [comment]
Another thing: I noticed that it often crashed in the evening long after I started it. I also noticed in the above log that I start and stop sound output quite a lot. Perhaps the leak was caused by that?

**Comment by wwmm** [comment]
I did a few tests starting and stopping the pipeline a few times and at least on gnome system monitor I did not see any change to the memory usage. If there is a leak in this operation it is so small that the pipeline would have to be restarted a really large number of times. The amount in your logs is not enough.

It seems we are back to the start =/ The only pointer we have is that the system is killing us. Otherwise that `killed` message would not have been shown. The question now is why it is deciding to do that.

**Comment by aorozsa** [comment]
I installed and ran `pulseeffects-git` yesterday in debug mode yesterday and didn't notice anything, however, I did start it later than usual. I just woke up 5 minutes ago and I'll be running the background service for the next couple of days to see if it crashes under normal use.

**Comment by aorozsa** [comment]
@wwmm I still experienced a crash. I'll now try running it in debug mode.

**Comment by aorozsa** [comment]
You also mentioned that the kernel is killing pulseeffects. Would you recommend I try switching to the LTS kernel?

Another thing I'm considering doing is only using my laptop for the next few days to test if my main computer's ageing hardware is the cause.

**Comment by wwmm** [comment]
I do not think it is worth the effort. The LTS kernel probably has the same system. But at the same time I have no idea why we are being killed if memory is not a problem.

**Comment by aorozsa** [comment]
@wwmm For the past few days, I've been using my laptop as if it were my main system. It's a HP Probook 450 G4, and the OS setup is basically the same as my main system: the latest version of Manjaro KDE (stable branch), and using Linux Kernel 5.6.16-1. When using my main system, PE would usually crash in the evening, and sometimes it would cash in the middle of the day. When using my laptop, I never experienced any crashes at all.

At this point, I think it's safe to say that my main system's practically-dust audio hardware is what's been making PE crash; I don't know why exactly, but at this point, I don't see any other possibility. I tried collecting a bunch of logs, running a bunch of tests to see if high CPU usage was the culprit, I installed a completely different instance of Manjaro (and my backup didn't work :( ), and I installed a newer version of PE, yet none of those fixed the issue on my main system. On my laptop, I was folding in the background, using KDE, and running the version of PE from Manjaro's repositories (4.7.2), and I never experienced a crash in those three days.

Unless you have anything else you'd like to add, I'll go ahead and close this issue.
